# Task 4.5 – Data Consistency Checks

This notebook performs consistency checks on the Instacart datasets.

**Objectives:**
- Identify mixed-type columns
- Find and handle missing values
- Detect and remove duplicate records

## Importing libraries

In [1]:
# 📦 Import libraries
import pandas as pd
import numpy as np
import os

## Load Data
I'll work with:
- `products.csv` (Original Data)
- `orders_wrangled.csv` (Prepared Data)

In [2]:
path = r'/Users/canancengel/A4_Instacart Basket Analysis/02_Data'
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'))
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))

In [3]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

## Mixed-Type Columns Check
Detecting if any column has inconsistent data types using a loop.

In [4]:
# Test for mixed-type columns in df_prods
for col in df_prods.columns:
    types = df_prods[col].map(type)
    if len(types.unique()) > 1:
        print(f"Mixed types found in column: {col}")

Mixed types found in column: product_name


In [5]:
# Descriptive statistics for df_ords
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### Interpretation of df_ords.describe()
The `order_number`, `order_hour_of_day`, and `days_since_prior_order` columns appear consistent.  
- `order_hour_of_day` ranges from 0 to 23, which matches the 24-hour clock.  
- `days_since_prior_order` has a max of 30.0, indicating no unusually large gaps between orders.  
- The presence of `NaN` values in `days_since_prior_order` is expected for new customers.

In [6]:
# Same test for df_ords
for col in df_ords.columns:
    types = df_ords[col].map(type)
    if len(types.unique()) > 1:
        print(f"Mixed types found in column: {col}")

There are no mixed-type columns detected in the `df_ords` dataframe. All columns are consistent in data type.

## Missing Values Check
I'll identify missing values in both dataframes.

In [7]:
import pandas as pd

In [8]:
df_prods = pd.read_csv('/Users/canancengel/A4_Instacart Basket Analysis/02_Data/Original Data/products.csv')

In [9]:
df_prods_clean = df_prods[df_prods['product_name'].notnull() & (df_prods['product_name'].str.strip() != "")]
df_prods_clean = df_prods_clean.drop_duplicates()
print(df_prods_clean.shape)

(49672, 5)


In [10]:
# Descriptive statistics for df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [11]:
df_ords = pd.read_csv('/Users/canancengel/A4_Instacart Basket Analysis/02_Data/Prepared Data/orders_wrangled.csv')

In [12]:
df_ords_clean = df_ords.drop_duplicates()

In [13]:
df_ords_clean['new_customer'] = df_ords_clean['days_since_prior_order'].isnull()

In [14]:
# Show number of missing values
df_ords_clean['days_since_prior_order'].isnull().sum()

206209

In [15]:
print(df_ords_clean.shape)

(3421083, 7)


In [16]:
# Number of missing values in days_since_prior_order
df_ords_clean['days_since_prior_order'].isnull().sum()

206209

The number of missing values in the `days_since_prior_order` column indicates the number of new customers. 
These missing values have not been removed from the dataset, as they provide important information.

In [17]:
# Descriptive statistics for df_ords
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

After identifying and removing missing values from the products dataframe and addressing missing values in the orders dataframe, I created clean versions of each dataset.

- The cleaned products dataframe contains 49,672 rows and 5 columns (rows with missing product names were removed).

- The cleaned orders dataframe contains 3,214,874 rows and 7 columns. Rows with missing days_since_prior_order were not removed as these represent new customers; instead, a new column (new_customer) was created to flag these orders.

These cleaned dataframes will be used for all subsequent analysis steps to ensure data consistency and accuracy.

## Handling Missing Product Names
I'll create a subset to view rows with missing product names.

In [18]:
# Subset for missing product names
df_nan = df_prods[df_prods['product_name'].isnull()]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Duplicates Check
I'll check for duplicate rows in both dataframes.

In [19]:
# Duplicates in df_prods
df_prods.duplicated().sum()

5

In [20]:
# Duplicates in df_ords
df_ords.duplicated().sum()

0

## Drop Duplicates

In [21]:
# Drop duplicates
df_prods = df_prods.drop_duplicates()
df_ords = df_ords.drop_duplicates()

## Final Checks Complete
Dataframes are now checked for consistency:
- Mixed types handled
- Missing values located
- 5 Duplicates removed from df_prods

## Export

## Exporting Cleaned Data

The cleaned products and orders dataframes have been exported as CSV files into the "Prepared Data" folder.  
These files are named `products_cleaned.csv` and `orders_cleaned.csv`.

In [23]:
df_prods_clean.to_csv('/Users/canancengel/A4_Instacart Basket Analysis/02_Data/Prepared Data/products_cleaned.csv', index=False)
df_ords_clean.to_csv('/Users/canancengel/A4_Instacart Basket Analysis/02_Data/Prepared Data/orders_cleaned.csv', index=False)